src/layers/python.rs

```diff
@@ -0,0 +1,582 @@
+                    format!("setuptools=={setuptools_version}").as_str(),
+                    format!("wheel=={wheel_version}").as_str(),
+                ])
+                .env_clear()
+                .envs(&command_env),
+        )
+        .map_err(PythonLayerError::BootstrapPipCommand)?;
+
+        // By default Pip will install into the system site-packages directory if it is writeable
+        // by the current user. Whilst the buildpack's own `pip install` invocations always use
+        // `--user` to ensure application dependencies are instead installed into the user
+        // site-packages, it's possible other buildpacks or custom scripts may forget to do so.
+        // By making the system site-packages directory read-only, Pip will automatically use
+        // user installs in such cases:
+        // https://github.com/pypa/pip/blob/23.0/src/pip/_internal/commands/install.py#L715-L773
+        fs::set_permissions(site_packages_dir, Permissions::from_mode(0o555))
+            .map_err(PythonLayerError::MakeSitePackagesReadOnlyIo)?;
+
+        let layer_metadata = self.generate_layer_metadata(&context.stack_id);
+        LayerResultBuilder::new(layer_metadata)
+            .env(layer_env)
+            .build()
+    }
+
+    fn existing_layer_strategy(
+        &self,
+        context: &BuildContext<Self::Buildpack>,
+        layer_data: &LayerData<Self::Metadata>,
+    ) -> Result<ExistingLayerStrategy, <Self::Buildpack as Buildpack>::Error> {
+        let cached_metadata = &layer_data.content_metadata.metadata;
+        let new_metadata = self.generate_layer_metadata(&context.stack_id);
+
+        if let Some(reason) = cache_invalidation_reason(cached_metadata, &new_metadata) {
+            log_info(format!("Discarding cache {reason}"));
+            Ok(ExistingLayerStrategy::Recreate)
+        } else {
+            log_info(format!(
+                "Using cached Python {}",
+                cached_metadata.python_version
+            ));
+            let PackagingToolVersions {
+                pip_version,
+                setuptools_version,
+                wheel_version,
+            } = &cached_metadata.packaging_tool_versions;
+            log_info(format!(
+                "Using cached pip {pip_version}, setuptools {setuptools_version} and wheel {wheel_version}"
+            ));
+            Ok(ExistingLayerStrategy::Keep)
+        }
+    }
+}
+
+impl<'a> PythonLayer<'a> {
+    fn generate_layer_metadata(&self, stack_id: &StackId) -> PythonLayerMetadata {
+        PythonLayerMetadata {
+            stack: stack_id.clone(),
+            python_version: self.python_version.to_string(),
+            packaging_tool_versions: self.packaging_tool_versions.clone(),
+        }
+    }
+}
+
+/// Metadata stored in the generated layer that allows future builds to determine whether
+/// the cached layer needs to be invalidated or not.
+#[derive(Clone, Deserialize, PartialEq, Serialize)]
```

Comment:
`Debug` all structs by default please.

```suggestion
#[derive(Debug, Clone, Deserialize, PartialEq, Serialize)]
```